File: go/stocks/portfolio.go
package stocks

type Portfolio []Money

func convert(money Money, currency string) float64 {
	eurToUsd := 1.2
	if money.currency == currency {
		return money.amount
	}
	return money.amount * eurToUsd
}

func (p Portfolio) Add(m Money) Portfolio {
	p = append(p, m)
	return p
}

func (p Portfolio) Evaluate(currency string) Money {
	total := 0.0
	for _, m := range p {
		total = total + convert(m, currency)
	}
	return Money{amount: total, currency: currency}
}

File: ts/src/stocks/portfolio.ts
import { Money } from "./Money";

export class Portfolio {
  private moneys: Money[];
  private eurToUsd: number
  constructor() {
    this.moneys = [];
    this.eurToUsd = 1.2
  }
  add(...moneys: Money[]) {
    this.moneys = this.moneys.concat(moneys);
  }
  convert(money: Money, currency: string) {
    if (money.currency === currency) {
      return money.amount;
    }
    return money.amount * this.eurToUsd;
  }
  evaluate(currency: string) {
    let total = this.moneys.reduce((sum, money) => {
      return sum + this.convert(money, currency);
    }, 0);
    return new Money(total, currency);
  }
}
